### PR TITLE
Texture coordinates and texture files in Scene

### PR DESCRIPTION
**Pull request by rs22**
This extends Scene, PolygonalGeometry and PolygonalDrawable in order to hold material/texture Information. It also extends the assimp loader to populate the Scene with this information, if available.

**Limitations/possible enhancements:**

 - Only the first texture coordinate set is loaded
 - Only the first texture file name per material is loaded, and only the diffuse texture is considered
 - No texture parameters are loaded (e.g. if image mapping is repeat, mirror,...)
 - Only the filename of the material's texture is stored in the Scene - a Material primitive class could be useful for more advanced use cases
 - Management of Texture-Objects is not integrated with PolygonalDrawable yet

For a demo on how to load and bind the textures using QImage see the GeometryStage.cpp in the massivelighting branch of https://github.com/valentin-pinkau/glexamples

It would be great if other groups could build on top of this :wink:

### CODE REVIEW

**Comment by sbusch42** [comment]
Looks very good so far!
But about the last comment: textures should *not* be loaded and used using QImage, because gloperate painters should not be dependent on Qt qt all. gloperate is designed to be independent of the windowing system, and there is a Qt as well as a GLFW-based viewer that can display a gloperate painter. All resource loading shall therefore be abstracted and not be dependent on, e.g., Qt. In short: Do not use specific libraries, use the abstracted loader system instead (the same goes for assimp).

To load textures, use the generic resource system of gloperate:
     m_texture = m_resourceManager.load<globjects::Texture>("data/texture.png");

The viewer has to make sure, that the appropriate loading backends are registered.

**Comment by rs22** [comment]
Cool - I didn't know that the ResourceManager also works for textures. To make our example work again, I had to transform the texture file extension to lowercase, otherwise the QtTextureLoader could not be resolved. The fix is now also included in this PR.

File: source/gloperate-assimp/source/AssimpSceneLoader.cpp [hunk 1]
```diff
@@ -178,9 +187,22 @@ PolygonalGeometry * AssimpSceneLoader::convertGeometry(const aiMesh * mesh) cons
...
     }
 
+	if (mesh->HasTextureCoords(0))
+	{
+		// Copy texture cooridinate array
+		std::vector<glm::vec3> textureCoordinates;
+		for (size_t i = 0; i < mesh->mNumVertices; ++i)
+		{
+			const auto & textureCoordinate = mesh->mTextureCoords[0][i];
+			textureCoordinates.push_back({ textureCoordinate.x, textureCoordinate.y, textureCoordinate.z });
```
Comment (scheibel) [inline source/gloperate-assimp/source/AssimpSceneLoader.cpp]: When using this call, you can also use ```emplace_back``` (http://en.cppreference.com/w/cpp/container/vector/emplace_back)

File: source/gloperate/include/gloperate/primitives/PolygonalGeometry.h [hunk 2]
```diff
@@ -126,11 +126,66 @@ class GLOPERATE_API PolygonalGeometry
...
+	unsigned int materialIndex() const;
+
+	/**
+	*  @brief
+	*    Set material index
+	*
+	*  @param[in] materialIndex
+	*    Material index
+	*/
+	void setMaterialIndex(const unsigned int materialIndex);
```
Comment (scheibel) [inline source/gloperate/include/gloperate/primitives/PolygonalGeometry.h]: For the declaration (mainly in the header file), the const for always copied types is both not required and sometimes punished with a warning. You should remove it from the declaration. But then again, you are encouraged to use the additional const in the definition of the function / method (this is actually possible, allowed, and serves the exact purpose I described here.

**Comment by scheibel** [comment]
Please adjust your code according my comments, then :shipit:

File: source/gloperate-assimp/source/AssimpSceneLoader.cpp (hunk 1 above)
Comment (mjendruk) [inline source/gloperate-assimp/source/AssimpSceneLoader.cpp]: No, he shouldn't, we had an issue before. `emplace_back` is probably not supported by gcc 4.7

File: source/gloperate/include/gloperate/primitives/PolygonalDrawable.h [hunk 3]
```diff
@@ -73,11 +73,12 @@ class GLOPERATE_API PolygonalDrawable
...
-    globjects::ref_ptr<globjects::VertexArray> m_vao;       /**< Vertex array object */
-    globjects::ref_ptr<globjects::Buffer>      m_indices;   /**< Index buffer */
-    globjects::ref_ptr<globjects::Buffer>      m_vertices;  /**< Vertex buffer */
-    globjects::ref_ptr<globjects::Buffer>      m_normals;   /**< Normal buffer (may be empty) */
-    gl::GLsizei                                m_size;      /**< Number of elements (m_indices) */
+    globjects::ref_ptr<globjects::VertexArray> m_vao;                 /**< Vertex array object */
+    globjects::ref_ptr<globjects::Buffer>      m_indices;             /**< Index buffer */
+    globjects::ref_ptr<globjects::Buffer>      m_vertices;            /**< Vertex buffer */
+    globjects::ref_ptr<globjects::Buffer>      m_normals;             /**< Normal buffer (may be empty) */
+	globjects::ref_ptr<globjects::Buffer>      m_textureCoordinates;  /**< Texture coordinate buffer (may be empty) */
```
Comment (mjendruk) [inline source/gloperate/include/gloperate/primitives/PolygonalDrawable.h]: replace with spaces please

File: source/gloperate-assimp/source/AssimpSceneLoader.cpp (hunk 1 above)
Comment (scheibel) [inline source/gloperate-assimp/source/AssimpSceneLoader.cpp]: I found this issue for ```map``` and ```set``` but not for ```vector```. Here is a blog entry claiming that it works with gcc 4.7: http://lemire.me/blog/archives/2012/06/20/do-not-waste-time-with-stl-vectors/

Comment (mjendruk) [inline source/gloperate-assimp/source/AssimpSceneLoader.cpp]: Okay, I'm not sure about it. Maybe @sbusch42 can tell us, if it works or not.

**Comment by rs22** [comment]
I fixed the method signature and replaced the tab with spaces.
I didn't change push_back to emplace_back in the Assimp loader - if it turns out to be better, we should probably adjust the vertices and normals import as well. Until then, the code is consistent at least.

**Comment by scheibel** [comment]
Not integrating ```emplace_back``` is totally fine, I mainly wanted you to show the actual method to support your use case :wink:

**Pull request merged**